search/v2/geocoder.cpp

```diff
@@ -242,35 +248,114 @@ void Geocoder::FillLocalitiesTable(MwmContext const & context)
+          context.m_value, static_cast<my::Cancellable const &>(*this), m_retrievalParams));
+  }
+
+  // 2. Get all locality candidates with all the token ranges.
+  vector<Locality> preLocalities;
+
+  for (size_t i = 0; i < m_numTokens; ++i)
+  {
+    CBVPtr intersection;
+    intersection.Set(tokensCBV[i].get(), false /*isOwner*/);
+    if (intersection.IsEmpty())
+      continue;
 
-    m2::PointD const center = feature::GetCenter(ft, FeatureType::WORST_GEOMETRY);
-    double const radiusM = ftypes::GetRadiusByPopulation(ft.GetPopulation());
+    for (size_t j = i + 1; j <= m_numTokens; ++j)
+    {
+      coding::CompressedBitVectorEnumerator::ForEach(*intersection.Get(),
+               [&](uint32_t featureId)
+               {
+                 Locality l;
+                 l.m_featureId = featureId;
+                 l.m_startToken = i;
+                 l.m_endToken = j;
+                 preLocalities.push_back(l);
+               });
+
+      if (j < m_numTokens)
+      {
+        intersection.Intersect(tokensCBV[j].get());
+        if (intersection.IsEmpty())
+          break;
+      }
+    }
+  }
 
-    Locality locality;
-    locality.m_featureId = featureId;
-    locality.m_startToken = curToken;
-    locality.m_endToken = endToken;
-    locality.m_rect = MercatorBounds::RectByCenterXYAndSizeInMeters(center, radiusM);
-    m_localities[make_pair(curToken, endToken)].push_back(locality);
+  auto const tokensCountFn = [&](Locality const & l)
+  {
+    // Important! Don't take into account matched prefix for locality comparison.
+    size_t d = l.m_endToken - l.m_startToken;
+    ASSERT_GREATER(d, 0, ());
+    if (l.m_endToken == m_numTokens && !m_params.m_prefixTokens.empty())
+      --d;
+    return d;
   };
 
-  for (size_t curToken = 0; curToken < m_numTokens; ++curToken)
+  // 3. Unique preLocalities with featureId but leave the longest range if equal.
+  sort(preLocalities.begin(), preLocalities.end(),
+        [&](Locality const & l1, Locality const & l2)
+        {
+          if (l1.m_featureId != l2.m_featureId)
+            return l1.m_featureId < l2.m_featureId;
+          return tokensCountFn(l1) > tokensCountFn(l2);
+        });
+
+  preLocalities.erase(unique(preLocalities.begin(), preLocalities.end(),
+                              [](Locality const & l1, Locality const & l2)
+                              {
+                                return l1.m_featureId == l2.m_featureId;
+                              }), preLocalities.end());
+
+  // 4. Leave most popular localities.
+  // Use 2*kMaxLocalitiesCount because there can be countries, states, ...
+  if (preLocalities.size() > 2*kMaxLocalitiesCount)
   {
-    for (size_t endToken = curToken + 1; endToken <= m_numTokens; ++endToken)
+    auto rankTable = search::RankTable::Load(context.m_value.m_cont);
+
+    sort(preLocalities.begin(), preLocalities.end(),
+          [&] (Locality const & l1, Locality const & l2)
```

Comment:
BTW, why not to create in the anonymous namespaces short helpers, like
```c++
struct CompareByTokensCountAndRank { ... }
struct CompareByIdAndTokensCount { ... }
struct SameFeatureId { ... }
```
The code here will be much cleaner.